### PR TITLE
MAINT: Fix the cutoff value inconsistency for pinv2 and pinvh

### LINKED ISSUE #8861: scipy.linalg.pinv gives wrong result while scipy.linalg.pinv2 works
For a student project with finite differences we use the pseudo inverse. Here, we noticed large differences in the accuracy of `scipy.linalg.pinv` and `scipy.linalg.pinv2`: while the latter works just fine, the first produces useless pseudo inverses in this example, see the code below.

As far as I understand, `pinv` uses `scipy.linalg.lstsq` to minimize | e^i - A x^i |_2 for the i-th unit vector e^i in order to obtain the pseudo inverse X, whereas `pinv` directly uses a truncated singular value decomposition.
However, internally `lstsq` uses the LAPACK routine `gelss` which solves the minimization also using a singular value decomposition, see e. g. http://www.netlib.org/lapack/lug/node27.html .

The reason why `pinv` fails for the example below while `pinv2` doesn't results from the fact, that the standard cut-off threshold for the small singular values in `pinv2` is set to `1e6*eps` (for double), whereas `pinv` uses `eps*max(n,m)` if `A` is an `n`x`m` matrix, which is much smaller for the matrix in our example.

Now I wonder why `pinv` and `pinv2` use different cut-off thresholds as standard values. If I use the standard value of `pinv2` for `pinv`, see comment in example code, the results are about the same.
Moreover, I wonder why the least squares solver is namend `pinv` and the *classical* method is called `pinv2`. In my opinion it might by better to switch them.
Since in the end, both use a truncated singular value decomposition, I do not completely understand the fundamental need of both.

### Reproducing code example:
```
import numpy as np
import scipy.sparse as sp
import scipy.linalg as la

ns = 4,
hs = [1 / (n + 1) for n in ns]

def get_A_h(n):
    T_diagonals = [[4] * (n + 2), [-2] + [-1] * n, [-1] * n + [-2]]
    T = sp.diags(T_diagonals, [0, 1, -1])   
    Ts_diagonals = [[-2] + [-1] * n, [-1] * n + [-2]]
    Ts = sp.diags(Ts_diagonals, [1, -1])    
    I_n2 = np.identity(n + 2)
    A_h = (n + 1) ** 2 * (sp.kron(I_n2, T) + sp.kron(Ts, I_n2))
    return A_h

def check_moore_penrose_axioms(A, Apinv):
    print()
    print("1:", la.norm(A@Apinv@A - A))
    print("2:", la.norm(Apinv@A@Apinv - Apinv))
    print("3:", la.norm(A@Apinv - (A@Apinv).T))
    print("4:", la.norm(Apinv@A - (Apinv@A).T))

for n in ns:
    N = n + 2

    A_h = get_A_h(n).toarray()
    R = np.eye(*A_h.shape)
    
    print(A_h)
    
    A_h_plus = la.pinv(A_h) #, rcond=1e6*np.finfo(A_h.dtype).eps)    
    check_moore_penrose_axioms(A_h, A_h_plus)
    
    A_h_plus = la.pinv2(A_h, )
    check_moore_penrose_axioms(A_h, A_h_plus)
```

### Output:
```
[[100. -50.   0. ...   0.   0.   0.]
 [-25. 100. -25. ...   0.   0.   0.]
 [  0. -25. 100. ...   0.   0.   0.]
 ...
 [  0.   0.   0. ... 100. -25.   0.]
 [  0.   0.   0. ... -25. 100. -25.]
 [  0.   0.   0. ...   0. -50. 100.]]

1: 10.966687766642307
2: 3108066184397.3657
3: 0.313565155535093
4: 2.715716258843298

1: 1.2034095577848958e-12
2: 4.613607271655997e-16
3: 2.7619937032685725e-14
4: 2.531755927761354e-14
```

### Scipy/Numpy/Python version information:
```
1.0.1 1.14.3 sys.version_info(major=3, minor=6, micro=5, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by ilayn**
Fixes #8861 

The least square solver bound value is notoriously difficult to select due to the inherent trade-off between selecting the best minimizer for the ||Ax-b|| and the precision of x when b is in the image of the matrix A. Since `pinv` is a special least squares problem the usual bound of dtype eps is relaxed to have a factor of `MAX(M, N)`. Otherwise the problem shown in the linked issue is hit more often than acceptable times.

On the other hand SVD based pseudo-inverses enjoy the convenience of having the singular values available for obtaining an upper bound on the error. In this PR we remove the hard-coded values of 1E3 and 1E6 factors for pinv2 and pinvh and tie them to the `max(M,N)*sigma_max*eps`. Probably for historical reasons the cutoff values were hardcoded. 

Again, these values by no means provide a failsafe situation. However, they are kind of the standard values that serve well in most of the cases. Otherwise one can always provide a better `cond` value manually.

Here a possible backwards incompatibility can arise if a user relies on the resulting precision too heavily. But in my opinion, it repairs quite more than it might break.

### CODE REVIEW

**Comment by rgommers** [comment]
Can you explain in the description in what cases this matters for backwards compatibility?

**Comment by tylerjereddy** [comment]
@ilayn If you are comfortable assessing `pinv` related issues, note that NumPy has recently been plagued by sporadic / bizzare `pinv` failures: https://github.com/numpy/numpy/issues/12862

Probably a little too optimistic to hope for any connection though.

**Comment by ilayn** [comment]
@rgommers Done. 

@tylerjereddy I'll have a look.

**Comment by ilayn** [comment]
@tylerjereddy It looks like the same issue fixed here. For single precision (np.float32, np.complex64 etc.) pinv value is still `1e-15` by default and that causes mismatches almost every failed test very close the single precision eps (~1e-7). 

If OpenBLAS folks touched these and their related LAPACK routines or number representations, they can flip to the left or to the right if they are sitting on the fence. Hence I am not so sure if it is compiler related. 

Also, unfortunately, `np.linalg.pinv` is `sp.linalg.pinv2` :(

**Comment by rgommers** [comment]
Thanks @ilayn, clear description. Let's leave this open for a while to give people a chance to weigh in.

**Comment by ilayn** [comment]
If there are no more comments on this, I'd like to merge it tomorrow.

File: scipy/linalg/basic.py [hunk 1]
```diff
@@ -1389,12 +1386,9 @@ def pinvh(a, cond=None, rcond=None, lower=True, return_rank=False,
...
     cond, rcond : float or None
-        Cutoff for 'small' singular values.
-        Singular values smaller than ``rcond*largest_singular_value``
-        are considered zero.
-        If None and the dtype of `a` is ``np.float32``, it is set to
-        ``np.finfo('float32').eps * 1e3``.
-        Otherwise, it is set to ``np.finfo('float64').eps * 1e6``.
+        Cutoff for 'small' singular values. If omitted, singular values smaller
+        than ``max(M,N)*largest_singular_value*eps`` are considered zero where
+        ``eps`` is the machine precision.
```
Comment (larsoner) [inline scipy/linalg/basic.py]: This could use a versionchanged directive

**Review by larsoner — COMMENTED** [review]
Looks good to me. Probably worth a note in the backward incompatible changes release notes section, too, though.

Also some tests that show things work better now than on master would help if it's not too much work. Maybe something simple like pinv, pinvh, and pinv2 (good case for parametrize) on a 2x2 double array with 1 and 1e-10 in the diagonal?

File: scipy/linalg/basic.py (hunk 1 above)
Comment (ilayn) [reply scipy/linalg/basic.py]: Will add.

**Comment by ilayn** [comment]
Ah yes. Good idea.

**Comment by ilayn** [comment]
@larsoner pinvh is basically pinv2 with different solver so I only compared pinv and pinv2. I couldn't replicate a symmetric matrix that exhibits the issue but let me know if you have a better example to test with

**Comment by ilayn** [comment]
Sigh... pypy is not having a good day again. @mattip Any suggestions?

**Comment by mattip** [comment]
This is using NumPy 1.16.3 I assume, which was released yesterday and added a new C-API function. I will issue a PR to go back to using nightly builds rather than official releases, since PyPy fixed when it was merged to NumPy but it is not in a release yet..

**Comment by mattip** [comment]
Maybe better would be to pin the NumPy used in the PyPy CI run to the version used in the other CI runs.

**Comment by ilayn** [comment]
@tylerjereddy I forgot to add the milestone to this if @larsoner is happy with the changes it is ok to merge.

File: scipy/linalg/basic.py [hunk 2]
```diff
@@ -1293,7 +1297,9 @@ def pinv(a, cond=None, rcond=None, return_rank=False, check_finite=True):
     """
     a = _asarray_validated(a, check_finite=check_finite)
     b = np.identity(a.shape[0], dtype=a.dtype)
-    if rcond is not None:
+    if rcond is None:
+        cond = max(a.shape) * np.spacing(a.real.dtype.type(1))
+    else:
         cond = rcond
```
Comment (larsoner) [inline scipy/linalg/basic.py]: If I do:
```
pinv(a, 0.1)
```
my `cond=0.1` value will end up not being used because this code path will enter the `if rcond is None` part of the branch, and overwrite `cond`, right?

If so, we need some more tests that would have caught this sort of thing. I'd do a `pytest.mark.parametrize` over `cond, rcond` pairs operating on an array that, if the specified condition is not respected, will not `np.allclose` with an expected result. To make a very simple example with `0.1` as the potential `cond/rcond` and the expected result `[[1, 0], [0, 0]]` when it is used properly (and `np.eye(2)` when it is not):
```
>>> from scipy.linalg import pinv
>>> import numpy as np
>>> x = np.array([[1, 0], [0, 0.01]])
>>> pinv(x)
array([[  1.,   0.],
       [  0., 100.]])
>>> pinv(x, 0.1)
array([[1., 0.],
       [0., 0.]])
```

**Review by larsoner — COMMENTED** [review]
> pinvh is basically pinv2 with different solver so I only compared pinv and pinv2. I couldn't replicate a symmetric matrix that exhibits the issue but let me know if you have a better example to test with

What I meant above about "a 2x2 double array with 1 and 1e-10 in the diagonal" is something like the following, which should fail on `master` for pinvh and pinv2 (and succeed for pinv) but all pass on your PR (at least when I tested it as a script, now generalized to the `pytest` case, it did I think):
```
@pytest.mark.parametrize('scale', (1e-20, 1., 1e20))
@pytest.mark.parametrize('pinv_', (pinv, pinvh, pinv2))
def test_auto_rcond(scale, pinv_):
    x = np.array([[1, 0], [0, 1e-10]]) * scale
    expected = np.diag(1. / np.diag(x))
    x_inv = pinv_(x)
    assert_allclose(x_inv, expected)
```

File: scipy/linalg/basic.py (hunk 2 above)
Comment (ilayn) [reply scipy/linalg/basic.py]: Yes indeed here I didn't touch the original flow but merely inverted it. To be honest, I am not entirely sure what these keywords meant to do. And also why there are two of them (cond and rcond) since one is supposed to be the `r`eciprocal of the other but in the code there is no such distinction. Hence I didn't want to change the behavior but we can actually change it in a separate PR now that you also noticed it. 

But your test idea is a good one. I will try to find something that fails on master.

Comment (larsoner) [reply scipy/linalg/basic.py]: > Yes indeed here I didn't touch the original flow but merely inverted it.

Assuming by this you mean that the rcond/cond conditional behavior is unchanged (i.e., only what actually gets computed is changed, when it's necessary to do so), I think that's not the case currently. On `master` the conditional implies a truth table like:

- `rcond is None`
  - `cond is None`: compute (presumably? Both remain `None` so I assume it's computed somewhere later)
  - `cond is not None`: use `cond`
- `rcond is not None`
  - `cond is None`: use `rcond`
  - `cond is not None`: use `rcond`

In other words, it was a "prefer rcond, and if neither is provided, compute" mode.

On this PR, this becomes:

- `rcond is None`
  - `cond is None`: compute
  - `cond is not None`: **compute**
- `rcond is not None`
  - `cond is None`: use `rcond`
  - `cond is not None`: use `rcond`

This becomes a "rcond is used if it's not None, else compute (cond is ignored)". In other words, the `rcond is None, cond is not None` path has been broken. Or do I miss something here?

Comment (larsoner) [reply scipy/linalg/basic.py]: (FWIW I usually see `rcond` as a *relative* (to max singular value) condition, and `cond` as an *absolute* condition, but that doesn't seem to be how it was used here.)

Comment (ilayn) [reply scipy/linalg/basic.py]: I don't know where it is actually converted. This is the current master code 

https://github.com/scipy/scipy/blob/master/scipy/linalg/basic.py#L1296

and there is no conversion done anywhere else since it is passed directly to `scipy.linalg.lstsq`. I am even more confused now 😃 I think we botched this up at some point.

Comment (ilayn) [reply scipy/linalg/basic.py]: Ah now I see what you mean.

**Comment by tylerjereddy** [comment]
I think we should bump the milestone--looks like there's still some active discussion / revision going on here.

**Comment by ilayn** [comment]
Give me an hour and it will be done :)

**Comment by ilayn** [comment]
@larsoner After having sufficient coffee I can give a better response: Now if the cond arg is given it is respected and we only interfere if both are None which is what this PR is fixing. 

Also thank you for tests indeed it fails on master and passes this PR. Hence please have a look. 

But I think the terminology is wrong in the inversion context; just as an example https://nl.mathworks.com/help/matlab/ref/rcond.html

File: scipy/linalg/basic.py [hunk 3]
```diff
@@ -1317,12 +1325,14 @@ def pinv2(a, cond=None, rcond=None, return_rank=False, check_finite=True):
     a : (M, N) array_like
         Matrix to be pseudo-inverted.
     cond, rcond : float or None
-        Cutoff for 'small' singular values.
-        Singular values smaller than ``rcond*largest_singular_value``
-        are considered zero.
```
Comment (larsoner) [inline scipy/linalg/basic.py]: This older wording actually seems more readable to me. As above, you have removed the description of how `rcond` is effectively used when it is provided, in favor of describing how `None` (the default) is handled, which I think is more clearly described in a second sentence.

File: scipy/linalg/basic.py [hunk 4]
```diff
@@ -1257,10 +1257,14 @@ def pinv(a, cond=None, rcond=None, return_rank=False, check_finite=True):
         Matrix to be pseudo-inverted.
     cond, rcond : float, optional
         Cutoff for 'small' singular values in the least-squares solver.
-        Singular values smaller than ``rcond * largest_singular_value``
-        are considered zero.
-        If None, it is set to ``np.finfo(a.dtype).eps``.
-        If `a` is an array of integers, it is set to ``np.finfo('float64').eps``.
+        Singular values smaller than ``max(M, N) * eps`` are considered zero
```
Comment (larsoner) [inline scipy/linalg/basic.py]: The ``max(M, N) * eps`` looks more like a description of the default, than a description of the behavior of the parameter. So instead of saying "Singular values smaller than ..." it should probably say "If cond and rcond are None, a value of max(M,N)*eps is used".

`cond` gets passed directly to [lstsq](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lstsq.html), where it describes this as:

> Cutoff for ‘small’ singular values; used to determine effective rank of a. Singular values smaller than rcond * largest_singular_value are considered zero.

This seems like a clearer wording of the behavior than what is in this PR. I'd prefer that this docstring say the same thing as `lstsq` (same goes for `pinv2` and `pinvh`). After describing how `rcond` is used, then it should mention the default max(M,N)*eps.

**Review by larsoner — APPROVED** [review]
Apologies for being a pain here, but I'm not sure the docstring rewordings help the problem here.

File: scipy/linalg/basic.py (hunk 4 above)
Comment (ilayn) [reply scipy/linalg/basic.py]: Indeed, I apparently cut too much including the `if None` part.

File: scipy/linalg/basic.py (hunk 3 above)
Comment (ilayn) [reply scipy/linalg/basic.py]: indeed.

File: scipy/linalg/basic.py [hunk 5]
```diff
@@ -1256,11 +1256,15 @@ def pinv(a, cond=None, rcond=None, return_rank=False, check_finite=True):
     a : (M, N) array_like
         Matrix to be pseudo-inverted.
     cond, rcond : float, optional
-        Cutoff for 'small' singular values in the least-squares solver.
-        Singular values smaller than ``rcond * largest_singular_value``
-        are considered zero.
-        If None, it is set to ``np.finfo(a.dtype).eps``.
-        If `a` is an array of integers, it is set to ``np.finfo('float64').eps``.
+        Cutoff for 'small' singular values; singular values smaller than this
+        value are considered zero. If both are omitted, the default
```
Comment (larsoner) [inline scipy/linalg/basic.py]: I'm still not convinced this is an improvement. The old doc makes it clear that the cutoff provided is relative to the maximum singular value:

> Singular values smaller than ``rcond * largest_singular_value``

The wording of the new one makes it sound like an absolute cutoff (i.e., in each singular value's actual value, not relative to the maximum singular value):

> singular values smaller than this value are considered zero

Comment (ilayn) [reply scipy/linalg/basic.py]: Because it is an absolute cutoff. We specifically use this value passed to lstsq and it overrides it. That was the problem we are fixing.

Comment (ilayn) [reply scipy/linalg/basic.py]: Well if we are to follow all the way to lstsq ok I know what you mean. Let me reword that part.

Comment (larsoner) [reply scipy/linalg/basic.py]: If it were an absolute cutoff (not relative to the max singular value) then the test that I wrote would fail for the scale factor of 1e-20, no? Maybe I am misusing terminology somehow, but it seems like it must be using a "relative" singular value cutoff.

And either way, the docstring in these should match the docstring in `lstsq`, which still uses this relative terminology.

Comment (ilayn) [reply scipy/linalg/basic.py]: No I see your point. It is just that the terminology is really confusing for me since these are quite nonstandard usage.

**Comment by ilayn** [comment]
OK I thought `[ci skip]` would stop these and not the CircleCI sorry about that,

**Review by larsoner — APPROVED** [review]
Thanks for iterating, wording LGTM now

**Comment by ilayn** [comment]
Thanks for keeping it sane. I really confused myself with this one 😃

**Comment by tylerjereddy** [comment]
I restored the milestone to 1.3.0 & did the same for the cognate issue. If this deserves a release note please feel free to add one on the wiki page.

**Comment by ilayn** [comment]
Thanks Tyler, wiki page edited.

**Comment by GillesOrban** [comment]
Dear member of scipy,

Using `pinv2` and wanted to fix the relative conditioning number with `rcond`, I saw in the code that `rcond` and `cond` have the same behavior - as far as I can tell. 
Not sure if this is indented.
My suggestion would be to replace in line 1372
`cond = rcond`
by
`cond = rcond * np.max(s)`

Thank you for supporting scipy,

Gilles

**Comment by ilayn** [comment]
`cond = rcond` only if it is not `None` otherwise it has the `max(s)` by default

**Comment by ogrisel** [comment]
> cond = rcond only if it is not None otherwise it has the max(s) by default

I agree with @GillesOrban. @ilayn by reading the lines: https://github.com/scipy/scipy/pull/10067/files#diff-8400a871af5e3ff38643aff36ff192650313e769cdfe80f57b02140def686ee1R1368-R1377 one can see that:

- if `rcond is not None` and `cond is None`, then `cond = rcond` and `rank = np.sum(s > cond)` or equivalently `rank = np.sum(s > rcond)`
- if `rcond` is None and `cond is not None`, then `rank = np.sum(s > cond)` directly.

Therefore it those two arguments mean the same thing in practice.

Another way to frame it: there is no way to pass `xcond` and have `rank = np.sum(s > xcond * np.max(s))` (I am not sure if such an `xcond` should actually be named `cond` or `rcond`).

Finally the docstring does not explain what is the (expected) difference between those two arguments.

**Comment by ilayn** [comment]
Arrrgh, this issue still haunts me 🤦 

@ogrisel I think these keywords were incorrect to start with, as you mentioned. I'll come back to that in a sec. 

However I think there is a discrepancy in that rcond is not a relative condition number as in `rtol` to `atol` but the reciprocal condition number. So the functionality you want to pass is essentially a `rtol`, based on the largest singular value and makes sense (That's basically what the default value is trying to do in this PR). `rcond` and `cond` were placed as the condition number wrt inversion which is essentially an `atol` for cutoff and that's where the hardcoded numbers were placed. 

Back to keywords, ideally we shouldn't have had these wrongly named keywords, shortcut together behind the scenes but I didn't touch it for backwards compatibility. However I do agree that the naming is confusing. And if you want to have the relative cutoff value I would really appreciate a new issue/feature request as here not many people will see it. And more people can chime in. Probably we should have gone for `rtol` and `atol` with better names. 

Hopefully this time we can fix it for good and be done with this.
